option optimize_for = LITE_RUNTIME;

// Arduino proto that specifies how data is sent to/from SiteWhere.
//
// NOTE: This proto has been converted to use fixed64 rather than float for doubles
//       since arduino does not support a real double datatype.
message SiteWhere {

	enum Command {REGISTER = 1; ACKNOWLEDGE = 2; DEVICELOCATION = 3; DEVICEALERT = 4; DEVICEMEASUREMENT = 5; } 
	
	// Header contains the command and originator.
	message Header {
		required Command command = 1;
		optional string originator = 2;
	}

	// Register a device with SiteWhere.
	message RegisterDevice {
		required string hardwareId = 1;
		required string specificationToken = 2;
	}

	// Acknowledge previous command.
	message Acknowledge {
		required string hardwareId = 1;
		optional string message = 2;
	}

	// Report device location.
	message DeviceLocation {
		required string hardwareId = 1;
		required fixed64 latitude = 2;
		required fixed64 longitude = 3;
		optional fixed64 elevation = 4;
		optional fixed64 eventDate = 5;
	}

	// Report device alert.
	message DeviceAlert {
		required string hardwareId = 1;
		required string alertType = 2;
		required string alertMessage = 3;
		optional fixed64 eventDate = 4;
	}
	
	// A single measurement.
	message Measurement {
		required string measurementId = 2;
		required fixed64 measurementValue = 3;
	}

	// Event message for device measurements.
	message DeviceMeasurements {
		required string hardwareId = 1;
		repeated Measurement measurement = 2;
		optional fixed64 eventDate = 3;
	}
}

// Encapsulates data received from SiteWhere.
message Device {

	enum Command {REGISTER_ACK = 1; } 
	
	// Header contains the command and originator.
	message Header {
		required Command command = 1;
		optional string originator = 2;
		optional string nestedPath = 3;
		optional string nestedSpec = 4;
	}

	// Enumeration of registation state responses.
	enum RegistrationAckState {NEW_REGISTRATION = 1; ALREADY_REGISTERED = 2; REGISTRATION_ERROR = 3;} 
	enum RegistrationAckError {INVALID_SPECIFICATION = 1; SITE_TOKEN_REQUIRED = 2; } 

	// Acknowledges registration.
	message RegistrationAck {
		required RegistrationAckState state = 1;
		optional RegistrationAckError errorType = 2;
		optional string errorMessage = 3;
	}
}